File: app/src/main/java/com/mancode/financetracker/database/converter/DateConverter.java
package com.mancode.financetracker.database.converter;

import android.arch.persistence.room.TypeConverter;
import android.util.Log;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Created by Manveru on 02.02.2018.
 */

public class DateConverter {

    private static final String TAG = DateConverter.class.getName();

    public static final String DATE_FORMAT_STRING = "yyyy-MM-dd";

    @TypeConverter
    public static Date toDate(String string) {
        if (string == null) return null;
        Date result = null;
        try {
            DateFormat df = new SimpleDateFormat(DATE_FORMAT_STRING, Locale.getDefault());
            result = df.parse(string);
        } catch (ParseException e) {
            Log.d(TAG, "Error when trying to convert string [" + string + "] to date");
        }
        return result;
    }

    @TypeConverter
    public static String toString(Date date) {
        return date == null ? null :
                new SimpleDateFormat(DATE_FORMAT_STRING, Locale.US).format(date);
    }
}

File: app/src/main/java/com/mancode/financetracker/database/converter/DateConverter.kt
package com.mancode.financetracker.database.converter

import android.util.Log
import androidx.room.TypeConverter
import org.threeten.bp.LocalDate
import org.threeten.bp.format.DateTimeParseException

/**
 * Created by Manveru on 02.02.2018.
 */

object DateConverter {

    private val TAG = DateConverter::class.java.name

    /**
     * Same as DateTimeFormatter.ISO_LOCAL_DATE
     */
    const val DATE_FORMAT_STRING = "yyyy-MM-dd"

    @TypeConverter
    @JvmStatic
    fun toDate(string: String?): LocalDate? {
        if (string == null || string.isEmpty()) return null
        var result: LocalDate? = null
        try {
            result = LocalDate.parse(string)
        } catch (e: DateTimeParseException) {
            Log.d(TAG, "Error when trying to convert string [$string] to date")
        }

        return result
    }

    @TypeConverter
    @JvmStatic
    fun toString(date: LocalDate?): String? {
        return date?.toString()
    }
}
